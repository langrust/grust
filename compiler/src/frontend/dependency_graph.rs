mod component;
mod expression;
mod file;
mod statement;
mod stream_expression;
