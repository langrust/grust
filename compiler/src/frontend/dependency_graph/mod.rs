use petgraph::graphmap::DiGraphMap;

use crate::common::label::Label;

mod contract;
mod expression;
mod file;
mod node;
mod stream_expression;
mod statement;

/// Add an egde to the graph.
///
/// If edge already exits then keep the edge with the lowest weight.
pub fn add_edge(
    graph: &mut DiGraphMap<usize, Label>,
    signal_id: usize,
    dependency_id: usize,
    label: Label,
) {
    let prev_label = graph.add_edge(signal_id, dependency_id, label.clone());
    match (prev_label, label) {
        (Some(Label::Weight(prev_weight)), Label::Weight(new_weight))
            if prev_weight < new_weight =>
        {
            graph.add_edge(signal_id, dependency_id, Label::Weight(prev_weight));
        }
        _ => (),
    }
}
