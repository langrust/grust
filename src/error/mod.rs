use codespan_reporting::diagnostic::{Diagnostic, Label};

use crate::ast::{
    location::Location,
    type_system::Type,
};

/// Compilation errors enumeration.
/// 
/// [Error] enumeration is used during the compilation to alert
/// the programmer of some errors in its LanGRust program.
/// 
/// # Example
/// ```rust
/// use grustine::error::Error;
/// use grustine::ast::location::Location;
/// 
/// let mut errors = vec![];
/// 
/// let name = String::from("unknown");
/// let location = Location::default();
/// 
/// let error = Error::UnknownElement { name, location };
/// errors.push(error);
/// ```
/// 
#[derive(Debug, PartialEq, Eq, Clone)]
pub enum Error {
    /// encountering an unknown element
    UnknownElement {
        /// the unknow identifier
        name: String,
        /// the error location
        location: Location,
    },
    /// incompatible application
    IncompatibleInputType {
        /// given type as input
        given_type: Type,
        /// expected type as input
        expected_type: Type,
        /// the error location
        location: Location,
    },
}
impl Error {
    /// Transform the error into a diagnostic.
    /// 
    /// This makes it possible to use [codespan_reporting] to print
    /// pretty errors.
    /// 
    /// # Example
    /// ```rust
    /// use grustine::error::Error;
    /// use grustine::ast::location::Location;
    /// use codespan_reporting::{
    ///     files::SimpleFiles,
    ///     term::termcolor::{StandardStream, ColorChoice},
    ///     term,
    /// };
    /// 
    /// let mut errors: Vec<Error> = vec![];
    /// let mut files = SimpleFiles::new();
    /// 
    /// let file_id = files.add("file_test.gr", "a code without x...");
    /// let name = String::from("x");
    /// let location = Location {
    ///     file_id,
    ///     range: 0..0,
    /// };
    /// 
    /// let error = Error::UnknownElement { name, location };
    /// errors.push(error);
    /// 
    /// let writer = StandardStream::stderr(ColorChoice::Always);
    /// let config = term::Config::default();
    /// for error in &errors {
    ///     let writer = &mut writer.lock();
    ///     let _ = term::emit(writer, &config, &files, &error.to_diagnostic());
    /// }
    /// ```
    pub fn to_diagnostic(&self) -> Diagnostic<usize> {
        match self {
            Error::UnknownElement { name, location } => Diagnostic::error()
                .with_message("unknown element")
                .with_labels(vec![
                    Label::primary(location.file_id, location.range.clone())
                        .with_message("unknown")
                ])
                .with_notes(vec![
                    format!("element '{name}' is not defined")
                ]
            ),
            Error::IncompatibleInputType { given_type, expected_type, location } => Diagnostic::error()
                .with_message("incompatible application")
                .with_labels(vec![
                    Label::primary(location.file_id, location.range.clone())
                        .with_message("wrong input type")
                ])
                .with_notes(vec![
                    format!("expected '{expected_type}' but '{given_type}' was given")
                ]
            ),
        }
    }
}
