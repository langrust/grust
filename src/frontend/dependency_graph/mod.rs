use petgraph::graphmap::DiGraphMap;

use crate::common::label::Label;

mod expression;
mod file;
mod node;
mod stream_expression;
mod term;

pub fn add_edge(
    graph: &mut DiGraphMap<usize, Label>,
    signal_id: usize,
    dependency_id: usize,
    label: Label,
) {
    let prev_label = graph.add_edge(signal_id, dependency_id, label.clone());
    match (prev_label, label) {
        (Some(Label::Weight(prev_weight)), Label::Weight(new_weight))
            if prev_weight < new_weight =>
        {
            graph.add_edge(signal_id, dependency_id, Label::Weight(prev_weight));
        }
        _ => (),
    }
}
