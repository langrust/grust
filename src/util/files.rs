use codespan_reporting::files;
use std::ops::Range;

/// A file that is backed by an `Arc<String>`.
#[derive(Debug, Clone)]
struct File {
    /// The name of the file.
    name: String,
    /// The source code of the file.
    source: String,
    /// The starting byte indices in the source code.
    line_starts: Vec<usize>,
}

impl File {
    fn line_start(&self, line_index: usize) -> Result<usize, files::Error> {
        use std::cmp::Ordering;

        match line_index.cmp(&self.line_starts.len()) {
            Ordering::Less => Ok(*self
                .line_starts
                .get(line_index)
                .expect("failed despite previous check")),
            Ordering::Equal => Ok(self.source.len()),
            Ordering::Greater => Err(files::Error::LineTooLarge {
                given: line_index,
                max: self.line_starts.len() - 1,
            }),
        }
    }
}

/// An opaque file identifier.
#[derive(Copy, Clone, PartialEq, Eq, Debug)]
pub struct FileId(i32);
impl FileId {
    /// Construct and return the default file identifier.
    pub fn default() -> Self {
        return FileId(-1);
    }
}

/// The files handler.
#[derive(Debug, Clone)]
pub struct Files {
    files: Vec<File>,
}

impl Files {
    /// Create a new files database.
    pub fn new() -> Files {
        Files { files: Vec::new() }
    }

    /// Add a file to the database, returning the handle that can be used to
    /// refer to it again.
    pub fn add(
        &mut self,
        name: impl Into<String>,
        source: impl Into<String>,
    ) -> Option<FileId> {
        let file_id = FileId(i32::try_from(self.files.len()).ok()?);
        let name = name.into();
        let source = source.into();
        let line_starts = files::line_starts(&source).collect();

        self.files.push(File {
            name,
            line_starts,
            source,
        });

        Some(file_id)
    }

    /// Get the file corresponding to the given id.
    fn get(&self, file_id: FileId) -> Result<&File, files::Error> {
        self.files
            .get(file_id.0 as usize)
            .ok_or(files::Error::FileMissing)
    }
}

impl<'files> files::Files<'files> for Files {
    type FileId = FileId;
    type Name = &'files str;
    type Source = &'files str;

    fn name(&self, file_id: FileId) -> Result<&str, files::Error> {
        Ok(self.get(file_id)?.name.as_ref())
    }

    fn source(&self, file_id: FileId) -> Result<&str, files::Error> {
        Ok(&self.get(file_id)?.source)
    }

    fn line_index(
        &self,
        file_id: FileId,
        byte_index: usize
    ) -> Result<usize, files::Error> {
        self.get(file_id)?
            .line_starts
            .binary_search(&byte_index)
            .or_else(|next_line| Ok(next_line - 1))
    }

    fn line_range(
        &self,
        file_id: FileId,
        line_index: usize,
    ) -> Result<Range<usize>, files::Error> {
        let file = self.get(file_id)?;
        let line_start = file.line_start(line_index)?;
        let next_line_start = file.line_start(line_index + 1)?;

        Ok(line_start..next_line_start)
    }
}