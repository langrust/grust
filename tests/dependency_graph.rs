use codespan_reporting::files::SimpleFiles;

use grustine::dependency_graph;
use grustine::error::display;

#[test]
fn dependency_graph_of_counter() {
    let mut files = SimpleFiles::new();

    let counter_id = files.add(
        "counter.gr",
        std::fs::read_to_string("tests/fixture/dependency_graph/success/counter.gr")
            .expect("unkown file"),
    );

    match dependency_graph(counter_id, &mut files) {
        Ok(file) => insta::assert_yaml_snapshot!(file),
        Err(errors) => display(&errors, &files),
    }
}

#[test]
fn dependency_graph_of_blinking() {
    let mut files = SimpleFiles::new();

    let blinking_id = files.add(
        "blinking.gr",
        std::fs::read_to_string("tests/fixture/dependency_graph/success/blinking.gr")
            .expect("unkown file"),
    );

    match dependency_graph(blinking_id, &mut files) {
        Ok(file) => insta::assert_yaml_snapshot!(file),
        Err(errors) => display(&errors, &files),
    }
}

#[test]
fn dependency_graph_of_button_management() {
    let mut files = SimpleFiles::new();

    let button_management_id = files.add(
        "button_management.gr",
        std::fs::read_to_string("tests/fixture/dependency_graph/success/button_management.gr")
            .expect("unkown file"),
    );

    match dependency_graph(button_management_id, &mut files) {
        Ok(file) => insta::assert_yaml_snapshot!(file),
        Err(errors) => display(&errors, &files),
    }
}

#[test]
fn dependency_graph_of_button_management_condition_match() {
    let mut files = SimpleFiles::new();

    let button_management_condition_match_id = files.add(
        "button_management_condition_match.gr",
        std::fs::read_to_string(
            "tests/fixture/dependency_graph/success/button_management_condition_match.gr",
        )
        .expect("unkown file"),
    );

    match dependency_graph(button_management_condition_match_id, &mut files) {
        Ok(file) => insta::assert_yaml_snapshot!(file),
        Err(errors) => display(&errors, &files),
    }
}

#[test]
fn dependency_graph_of_button_management_using_function() {
    let mut files = SimpleFiles::new();

    let button_management_using_function_id = files.add(
        "button_management_using_function.gr",
        std::fs::read_to_string(
            "tests/fixture/dependency_graph/success/button_management_using_function.gr",
        )
        .expect("unkown file"),
    );

    match dependency_graph(button_management_using_function_id, &mut files) {
        Ok(file) => insta::assert_yaml_snapshot!(file),
        Err(errors) => display(&errors, &files),
    }
}

#[test]
fn dependency_graph_of_pid() {
    let mut files = SimpleFiles::new();

    let pid_id = files.add(
        "pid.gr",
        std::fs::read_to_string("tests/fixture/dependency_graph/success/pid.gr")
            .expect("unkown file"),
    );

    match dependency_graph(pid_id, &mut files) {
        Ok(file) => insta::assert_yaml_snapshot!(file),
        Err(errors) => display(&errors, &files),
    }
}
