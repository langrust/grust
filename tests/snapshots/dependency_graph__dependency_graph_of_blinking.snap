---
source: tests/dependency_graph.rs
expression: file
---
typedefs: []
functions:
  - id: 19
    statements: []
    returned:
      kind:
        Application:
          function_expression:
            kind:
              Identifier:
                id: 15
            typing:
              Abstract:
                - - Boolean
                  - Boolean
                  - Boolean
                - Boolean
            location: ~
            dependencies: ~
          inputs:
            - kind:
                Identifier:
                  id: 17
              typing: Boolean
              location: ~
              dependencies: ~
            - kind:
                Application:
                  function_expression:
                    kind:
                      Identifier:
                        id: 1
                    typing:
                      Abstract:
                        - - Boolean
                        - Boolean
                    location: ~
                    dependencies: ~
                  inputs:
                    - kind:
                        Identifier:
                          id: 18
                      typing: Boolean
                      location: ~
                      dependencies: ~
              typing: Boolean
              location: ~
              dependencies: ~
            - kind:
                Identifier:
                  id: 18
              typing: Boolean
              location: ~
              dependencies: ~
      typing: Boolean
      location: ~
      dependencies: ~
    location: ~
nodes:
  - id: 24
    unscheduled_equations:
      22:
        id: 22
        expression:
          kind:
            Expression:
              expression:
                Application:
                  function_expression:
                    kind:
                      Expression:
                        expression:
                          Identifier:
                            id: 15
                    typing:
                      Abstract:
                        - - Boolean
                          - Integer
                          - Integer
                        - Integer
                    location: ~
                    dependencies: []
                  inputs:
                    - kind:
                        Expression:
                          expression:
                            Identifier:
                              id: 20
                      typing: Boolean
                      location: ~
                      dependencies:
                        - - 20
                          - Weight: 0
                    - kind:
                        Expression:
                          expression:
                            Constant:
                              constant:
                                Integer: 0
                      typing: Integer
                      location: ~
                      dependencies: []
                    - kind:
                        Expression:
                          expression:
                            Application:
                              function_expression:
                                kind:
                                  Expression:
                                    expression:
                                      Identifier:
                                        id: 5
                                typing:
                                  Abstract:
                                    - - Integer
                                      - Integer
                                    - Integer
                                location: ~
                                dependencies: []
                              inputs:
                                - kind:
                                    Expression:
                                      expression:
                                        Application:
                                          function_expression:
                                            kind:
                                              Expression:
                                                expression:
                                                  Identifier:
                                                    id: 2
                                            typing:
                                              Abstract:
                                                - - Integer
                                                - Integer
                                            location: ~
                                            dependencies: []
                                          inputs:
                                            - kind:
                                                FollowedBy:
                                                  constant:
                                                    kind:
                                                      Expression:
                                                        expression:
                                                          Constant:
                                                            constant:
                                                              Integer: 0
                                                    typing: Integer
                                                    location: ~
                                                    dependencies: []
                                                  expression:
                                                    kind:
                                                      Expression:
                                                        expression:
                                                          Identifier:
                                                            id: 22
                                                    typing: Integer
                                                    location: ~
                                                    dependencies:
                                                      - - 22
                                                        - Weight: 0
                                              typing: Integer
                                              location: ~
                                              dependencies:
                                                - - 22
                                                  - Weight: 1
                                  typing: Integer
                                  location: ~
                                  dependencies:
                                    - - 22
                                      - Weight: 1
                                - kind:
                                    Expression:
                                      expression:
                                        Identifier:
                                          id: 23
                                  typing: Integer
                                  location: ~
                                  dependencies:
                                    - - 23
                                      - Weight: 0
                      typing: Integer
                      location: ~
                      dependencies:
                        - - 22
                          - Weight: 1
                        - - 23
                          - Weight: 0
          typing: Integer
          location: ~
          dependencies:
            - - 20
              - Weight: 0
            - - 22
              - Weight: 1
            - - 23
              - Weight: 0
        location: ~
      23:
        id: 23
        expression:
          kind:
            Expression:
              expression:
                Application:
                  function_expression:
                    kind:
                      Expression:
                        expression:
                          Identifier:
                            id: 15
                    typing:
                      Abstract:
                        - - Boolean
                          - Integer
                          - Integer
                        - Integer
                    location: ~
                    dependencies: []
                  inputs:
                    - kind:
                        Expression:
                          expression:
                            Identifier:
                              id: 21
                      typing: Boolean
                      location: ~
                      dependencies:
                        - - 21
                          - Weight: 0
                    - kind:
                        Expression:
                          expression:
                            Constant:
                              constant:
                                Integer: 1
                      typing: Integer
                      location: ~
                      dependencies: []
                    - kind:
                        Expression:
                          expression:
                            Constant:
                              constant:
                                Integer: 0
                      typing: Integer
                      location: ~
                      dependencies: []
          typing: Integer
          location: ~
          dependencies:
            - - 21
              - Weight: 0
        location: ~
    unitary_nodes: {}
    contract:
      requires: []
      ensures: []
      invariant: []
    location: ~
    graph:
      ? - 22
        - 20
      : Weight: 0
      ? - 22
        - 22
      : Weight: 1
      ? - 22
        - 23
      : Weight: 0
      ? - 23
        - 21
      : Weight: 0
  - id: 30
    unscheduled_equations:
      26:
        id: 26
        expression:
          kind:
            Expression:
              expression:
                Application:
                  function_expression:
                    kind:
                      Expression:
                        expression:
                          Identifier:
                            id: 15
                    typing:
                      Abstract:
                        - - Boolean
                          - Integer
                          - Integer
                        - Integer
                    location: ~
                    dependencies: []
                  inputs:
                    - kind:
                        Expression:
                          expression:
                            Identifier:
                              id: 29
                      typing: Boolean
                      location: ~
                      dependencies:
                        - - 29
                          - Weight: 0
                    - kind:
                        Expression:
                          expression:
                            Application:
                              function_expression:
                                kind:
                                  Expression:
                                    expression:
                                      Identifier:
                                        id: 5
                                typing:
                                  Abstract:
                                    - - Integer
                                      - Integer
                                    - Integer
                                location: ~
                                dependencies: []
                              inputs:
                                - kind:
                                    Expression:
                                      expression:
                                        Identifier:
                                          id: 27
                                  typing: Integer
                                  location: ~
                                  dependencies:
                                    - - 27
                                      - Weight: 0
                                - kind:
                                    Expression:
                                      expression:
                                        Constant:
                                          constant:
                                            Integer: 1
                                  typing: Integer
                                  location: ~
                                  dependencies: []
                      typing: Integer
                      location: ~
                      dependencies:
                        - - 27
                          - Weight: 0
                    - kind:
                        Expression:
                          expression:
                            Constant:
                              constant:
                                Integer: 0
                      typing: Integer
                      location: ~
                      dependencies: []
          typing: Integer
          location: ~
          dependencies:
            - - 29
              - Weight: 0
            - - 27
              - Weight: 0
        location: ~
      27:
        id: 27
        expression:
          kind:
            NodeApplication:
              node_id: 24
              inputs:
                - - 20
                  - kind:
                      Expression:
                        expression:
                          Identifier:
                            id: 28
                    typing: Boolean
                    location: ~
                    dependencies:
                      - - 28
                        - Weight: 0
                - - 21
                  - kind:
                      Expression:
                        expression:
                          Constant:
                            constant:
                              Boolean: true
                    typing: Boolean
                    location: ~
                    dependencies: []
              output_id: 22
          typing: Integer
          location: ~
          dependencies:
            - - 28
              - Weight: 0
        location: ~
      28:
        id: 28
        expression:
          kind:
            FollowedBy:
              constant:
                kind:
                  Expression:
                    expression:
                      Constant:
                        constant:
                          Boolean: true
                typing: Boolean
                location: ~
                dependencies: []
              expression:
                kind:
                  Expression:
                    expression:
                      Application:
                        function_expression:
                          kind:
                            Expression:
                              expression:
                                Identifier:
                                  id: 2
                          typing:
                            Abstract:
                              - - Boolean
                              - Boolean
                          location: ~
                          dependencies: []
                        inputs:
                          - kind:
                              Expression:
                                expression:
                                  Application:
                                    function_expression:
                                      kind:
                                        Expression:
                                          expression:
                                            Identifier:
                                              id: 9
                                      typing:
                                        Abstract:
                                          - - Integer
                                            - Integer
                                          - Boolean
                                      location: ~
                                      dependencies: []
                                    inputs:
                                      - kind:
                                          Expression:
                                            expression:
                                              Application:
                                                function_expression:
                                                  kind:
                                                    Expression:
                                                      expression:
                                                        Identifier:
                                                          id: 5
                                                  typing:
                                                    Abstract:
                                                      - - Integer
                                                        - Integer
                                                      - Integer
                                                  location: ~
                                                  dependencies: []
                                                inputs:
                                                  - kind:
                                                      Expression:
                                                        expression:
                                                          Identifier:
                                                            id: 27
                                                    typing: Integer
                                                    location: ~
                                                    dependencies:
                                                      - - 27
                                                        - Weight: 0
                                                  - kind:
                                                      Expression:
                                                        expression:
                                                          Constant:
                                                            constant:
                                                              Integer: 1
                                                    typing: Integer
                                                    location: ~
                                                    dependencies: []
                                        typing: Integer
                                        location: ~
                                        dependencies:
                                          - - 27
                                            - Weight: 0
                                      - kind:
                                          Expression:
                                            expression:
                                              Identifier:
                                                id: 25
                                        typing: Integer
                                        location: ~
                                        dependencies:
                                          - - 25
                                            - Weight: 0
                            typing: Boolean
                            location: ~
                            dependencies:
                              - - 27
                                - Weight: 0
                              - - 25
                                - Weight: 0
                typing: Boolean
                location: ~
                dependencies:
                  - - 27
                    - Weight: 0
                  - - 25
                    - Weight: 0
          typing: Boolean
          location: ~
          dependencies:
            - - 27
              - Weight: 1
            - - 25
              - Weight: 1
        location: ~
      29:
        id: 29
        expression:
          kind:
            Expression:
              expression:
                Application:
                  function_expression:
                    kind:
                      Expression:
                        expression:
                          Identifier:
                            id: 19
                    typing:
                      Abstract:
                        - - Boolean
                          - Boolean
                        - Boolean
                    location: ~
                    dependencies: []
                  inputs:
                    - kind:
                        Expression:
                          expression:
                            Identifier:
                              id: 28
                      typing: Boolean
                      location: ~
                      dependencies:
                        - - 28
                          - Weight: 0
                    - kind:
                        FollowedBy:
                          constant:
                            kind:
                              Expression:
                                expression:
                                  Constant:
                                    constant:
                                      Boolean: true
                            typing: Boolean
                            location: ~
                            dependencies: []
                          expression:
                            kind:
                              Expression:
                                expression:
                                  Identifier:
                                    id: 29
                            typing: Boolean
                            location: ~
                            dependencies:
                              - - 29
                                - Weight: 0
                      typing: Boolean
                      location: ~
                      dependencies:
                        - - 29
                          - Weight: 1
          typing: Boolean
          location: ~
          dependencies:
            - - 28
              - Weight: 0
            - - 29
              - Weight: 1
        location: ~
    unitary_nodes: {}
    contract:
      requires: []
      ensures: []
      invariant: []
    location: ~
    graph:
      ? - 26
        - 27
      : Weight: 0
      ? - 26
        - 29
      : Weight: 0
      ? - 27
        - 28
      : Weight: 0
      ? - 28
        - 25
      : Weight: 1
      ? - 28
        - 27
      : Weight: 1
      ? - 29
        - 28
      : Weight: 0
      ? - 29
        - 29
      : Weight: 1
component: ~
interface: ~
location: ~
